packages/prover/src/proof_provider/proof_provider.ts

```diff
@@ -0,0 +1,205 @@
+import {isForkExecution} from "@lodestar/params";
+import {allForks, capella} from "@lodestar/types";
+import {MAX_REQUEST_LIGHT_CLIENT_UPDATES} from "../constants.js";
+import {LightNode, RootProviderOptions as RootProviderInitOptions} from "../interfaces.js";
+import {assertLightClient} from "../utils/assertion.js";
+import {getExecutionPayloads, getGenesisData, getSlotLimitsForPayloads, getSyncCheckpoint} from "../utils/consensus.js";
+import {numberToHex} from "../utils/conversion.js";
+import {OrderedMap} from "./ordered_map.js";
+
+type RootProviderOptions = RootProviderInitOptions & {
+  transport: LightClientRestTransport;
+  api: Api;
+  config: ChainForkConfig;
+};
+
+export class ProofProvider {
+  private payloads: OrderedMap<string, allForks.ExecutionPayload> = new OrderedMap();
+  private finalizedSlot = 0;
+  private readyPromise?: Promise<void>;
+
+  // Map to match CL slot to EL block number
+  private slotsMap: {[slot: number]: string} = {};
+
+  lightClient?: Lightclient;
+
+  constructor(private options: RootProviderOptions) {}
+
+  async waitToBeReady(): Promise<void> {
+    return this.readyPromise;
+  }
+
+  static init(opts: RootProviderInitOptions): ProofProvider {
+    if (opts.mode === LightNode.P2P) {
+      throw new Error("P2P mode not supported yet");
+    }
+
+    const config = createChainForkConfig(networksChainConfig[opts.network]);
+    const api = getClient({urls: opts.urls}, {config});
+    const transport = new LightClientRestTransport(api);
+
+    const provider = new ProofProvider({
+      ...opts,
+      config,
+      api,
+      transport,
+    });
+
+    provider.readyPromise = provider.sync(opts.checkpoint);
+
+    return provider;
+  }
+
+  private async sync(checkpoint?: string): Promise<void> {
+    if (this.lightClient !== undefined) {
+      throw Error("Light client already initialized and syncing.");
+    }
+
+    const {api, config, transport} = this.options;
+    const checkpointRoot = await getSyncCheckpoint(api, checkpoint);
+    const genesisData = await getGenesisData(api);
+
+    this.lightClient = await Lightclient.initializeFromCheckpointRoot({
+      checkpointRoot,
+      config,
+      transport,
+      genesisData,
+    });
+
+    assertLightClient(this.lightClient);
+    this.lightClient.start();
+    this.registerEvents();
+
+    const headSlot = this.lightClient.getHead().beacon.slot;
+    await this.lightClient?.sync(
+      computeSyncPeriodAtSlot(headSlot - MAX_REQUEST_LIGHT_CLIENT_UPDATES),
+      computeSyncPeriodAtSlot(headSlot)
+    );
+
+    // Load the payloads from the CL
+    const {start, end} = await getSlotLimitsForPayloads(this.lightClient);
+    const payloads = await getExecutionPayloads(this.options.api, start, end);
+    for (const [slot, payload] of Object.entries(payloads)) {
+      const blockNumber = numberToHex(payload.blockNumber);
+      this.payloads.set(blockNumber, payload);
+      this.slotsMap[parseInt(slot)] = blockNumber;
```

Comment:
i don't understand slot to number mapping, all numbers <= finalized's number are finalized, all numbers <= head's number are present

if there is a mapping that needs to be stored is beacon block root => execution block root for the non-finalized set and tracking what is head block